### PR TITLE
AYS-1142 | Persisting the last selected institution in the database

### PR DESCRIPTION
**Pull request by MenekseYuncu**
### Checklist

Before submitting your pull request, ensure the following:

- [x] **Title and Branch Naming Conventions:**
    - The pull request title follows the
      standard: [Pull Request Naming Conventions](https://github.com/afet-yonetim-sistemi/ays-be/blob/main/CONTRIBUTING.md#pull-request-naming-conventions).
    - The branch name follows one of the conventions outlined in
      the [Branch Naming Conventions](https://github.com/afet-yonetim-sistemi/ays-be/blob/main/CONTRIBUTING.md#branch-naming-conventions).

- [x] **Local Testing:**
    - I have tested my changes locally on Postman, and they are working as expected.

- [x] **Code Quality:**
    - The code is formatted according to the project's coding guidelines and style.
    - The code has been reviewed to ensure its quality.
    - The code does not contain any issues flagged by SonarLint.

- [x] **Documentation:**
    - Necessary documentation has been added or existing documentation has been updated, specifically detailing changes made in Postman.
    - If a new error or exception type has been introduced, it has been added to the [Exception Management Analysis](https://afetyonetimsistemi.atlassian.net/wiki/x/MwDLC) document.
    - I have verified that the error code, description, and scenario are consistent with existing standards.

- [x] **Testing:**
    - Relevant unit tests have been written and included.
    - Relevant integration tests have been written and included.

- [x] **Reviewers and Assignees:**
    - Default reviewers have been assigned to this pull request.
    - Assignees have been added if necessary.

- [x] **Labels and Associations:**
    - No specific actions are required in the Labels and Associations section for this pull request.

### CODE REVIEW

File: src/main/resources/db/changelog/changes/202606212332-ays-ddl.xml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    <changeSet id="202606121516-ays-ddl" author="ays">
+        <!-- ============================ -->
+        <!-- DDL of AYS_USER              -->
+        <!-- ============================ -->
+
+        <addColumn tableName="AYS_USER">
+            <column name="LAST_SELECTED_INSTITUTION_ID" type="VARCHAR(36)">
+                <constraints nullable="true"/>
+            </column>
+        </addColumn>
```
Comment (agitrubard) [inline src/main/resources/db/changelog/changes/202606212332-ays-ddl.xml:13]: @MenekseYuncu Login Attempt üzerinden yönetmemiz gerekiyor aslında, oraya `LAST_INSTITUTION_ID` olarak ekleyebiliriz. Bu sayede giriş yaparken oradan okunur

**Review by agitrubard — CHANGES_REQUESTED** [review]

File: src/main/resources/db/changelog/changes/202606212332-ays-ddl.xml (hunk 1 above)
Comment (agitrubard) [reply src/main/resources/db/changelog/changes/202606212332-ays-ddl.xml:13]: Ek olarak Institution'ın token içerisine eklenmiş olması gerekiyor ki, onun üzerinden hangi institution için girildiğini bilebilelim. Eğer institution seçilmemişse, kullanıcıların giriş yapmalarını engellememek için, bir tanesini seçip giriş yapmasını sağlayalım şu an için

File: src/main/java/org/ays/auth/service/impl/AysAuthServiceImpl.java [hunk 2]
```diff
@@ -89,6 +89,15 @@ public AysToken authenticate(final AysLoginRequest loginRequest) {
                             user.setLoginAttempt(AysUser.LoginAttempt.builder().build());
                             user.getLoginAttempt().success();
                         });
+        
+        if (user.getLoginAttempt().getLastSelectedInstitutionId() == null) {
+            final String firstInstitutionId = user.getInstitutions().stream()
+                    .findFirst()
+                    .map(Institution::getId)
+                    .orElse(null);
```
Comment (agitrubard) [inline src/main/java/org/ays/auth/service/impl/AysAuthServiceImpl.java]: Null setlersek institution seçmemiş olacak ve bu şekilde giriş yapamamalı aslında, burayı tekrar değerlendirebilir miyiz?

File: src/main/java/org/ays/auth/service/impl/AysAuthServiceImpl.java [hunk 3]
```diff
@@ -89,6 +89,15 @@ public AysToken authenticate(final AysLoginRequest loginRequest) {
                             user.setLoginAttempt(AysUser.LoginAttempt.builder().build());
                             user.getLoginAttempt().success();
                         });
+        
+        if (user.getLoginAttempt().getLastSelectedInstitutionId() == null) {
```
Comment (agitrubard) [inline src/main/java/org/ays/auth/service/impl/AysAuthServiceImpl.java]:
```suggestion
        if (user.getLoginAttempt().hasNotLastSelectedInstitutionId()) {
```

File: src/main/java/org/ays/auth/model/AysUser.java [hunk 4]
```diff
@@ -232,6 +234,12 @@ public Claims getClaims() {
             claimsBuilder.add(AysTokenClaims.USER_LAST_LOGIN_AT.getValue(), this.loginAttempt.lastLoginAt.toString());
         }
 
+        if (this.loginAttempt != null && this.loginAttempt.lastSelectedInstitutionId != null) {
+            claimsBuilder.add(AysTokenClaims.LAST_SELECTED_INSTITUTION_ID.getValue(), this.loginAttempt.lastSelectedInstitutionId);
+        } else {
+            claimsBuilder.add(AysTokenClaims.LAST_SELECTED_INSTITUTION_ID.getValue(), currentInstitution.getId());
+        }
```
Comment (agitrubard) [inline src/main/java/org/ays/auth/model/AysUser.java]: Aslında buradaki `currentInstitution`, mantıken `lastSelectedInstitutionId` ile eşleşen institution olmalı. Bu yüzden burada ekstra bir "var mı yok mu" kontrolü yapmak yerine, `currentInstitutionId`'yi (selected institution'a göre) baştan belirleyip, `LAST_SELECTED_INSTITUTION_ID` claim'ine de ihtiyaç kalmayabilir.

`this.getInstitution(this.loginAttempt.lastSelectedInstitutionId);` şeklinde yapabiliriz doğru instituion'ı seçebilmek için

File: src/main/java/org/ays/auth/model/enums/AysTokenClaims.java [hunk 5]
```diff
@@ -17,6 +17,7 @@ public enum AysTokenClaims {
     USER_FIRST_NAME("userFirstName"),
     USER_LAST_NAME("userLastName"),
     USER_LAST_LOGIN_AT("userLastLoginAt"),
+    LAST_SELECTED_INSTITUTION_ID("lastSelectedInstitutionId"),
```
Comment (agitrubard) [inline src/main/java/org/ays/auth/model/enums/AysTokenClaims.java]: `AysUser` sınıfına yazdığım yorum sonrası buna ihtiyaç kalmayabilir
```suggestion
```

File: src/main/resources/db/changelog/changes/202606212332-ays-ddl.xml [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+    <changeSet id="202606212332-ays-ddl" author="ays">
+        <!-- ============================ -->
+        <!-- DDL of AYS_USER              -->
+        <!-- ============================ -->
+
+        <addColumn tableName="AYS_USER_LOGIN_ATTEMPT">
+            <column name="LAST_SELECTED_INSTITUTION_ID" type="VARCHAR(36)">
+                <constraints nullable="true"/>
+            </column>
```
Comment (agitrubard) [inline src/main/resources/db/changelog/changes/202606212332-ays-ddl.xml]:
```suggestion
            <column name="LAST_SELECTED_INSTITUTION_ID" type="VARCHAR(36)"/>
```

File: src/main/resources/db/changelog/changes/202606212332-ays-ddl.xml [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+        </addColumn>
+
+        <!-- ============================ -->
+        <!-- DDL of AYS_USER              -->
+        <!-- ============================ -->
```
Comment (agitrubard) [inline src/main/resources/db/changelog/changes/202606212332-ays-ddl.xml]:
```suggestion
        </addColumn>
        <!-- =============== -->
        <!-- DDL of AYS_USER -->
        <!-- =============== -->
```

File: src/main/resources/db/changelog/changes/202606212332-ays-ddl.xml [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog
+                   http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-4.20.xsd">
+
+    <changeSet id="202606212332-ays-ddl" author="ays">
+        <!-- ============================ -->
+        <!-- DDL of AYS_USER              -->
+        <!-- ============================ -->
+
+        <addColumn tableName="AYS_USER_LOGIN_ATTEMPT">
```
Comment (agitrubard) [inline src/main/resources/db/changelog/changes/202606212332-ays-ddl.xml]:
```suggestion
        <!-- =============== -->
        <!-- DDL of AYS_USER -->
        <!-- =============== -->
        <addColumn tableName="AYS_USER_LOGIN_ATTEMPT">
```

**Review by agitrubard — CHANGES_REQUESTED** [review]

File: src/main/java/org/ays/auth/service/impl/AysAuthServiceImpl.java (hunk 2 above)
Comment (MenekseYuncu) [reply src/main/java/org/ays/auth/service/impl/AysAuthServiceImpl.java]: @agitrubard selam, ilk kez giriş yapan ve henüz kurum seçimi bulunmayan ya da birden fazla kuruma sahip kullanıcılar için nasıl bir mantık kurgulamalıyız?

Birden fazla kurumu olanlarda findFirst() ile ilk aktif kurumu seçmeyi düşündüm. Fakat tanımlı hiç kurumu olmayan kullanıcılar için nasıl bir aksiyon almalıyız?

Comment (agitrubard) [reply src/main/java/org/ays/auth/service/impl/AysAuthServiceImpl.java]: @MenekseYuncu selam, kullanıcının kurumu yoksa zaten giriş yapamıyor olması gerekiyor. Bu sebeple de burada null setlemek yerine burada bulmasını sağlamalıyız. Eğer buraya gelene kadar kurumunun var olup olmadığı kontrol edilmemişse `orElseThrow()` ile hata fırlatabiliriz. Giriş kısmında ise gerekli kontrolleri ekleyebiliriz bu task ile birlikte.

**Review by agitrubard — CHANGES_REQUESTED** [review]

File: src/main/java/org/ays/auth/model/AysUser.java [hunk 9]
```diff
@@ -199,13 +201,49 @@ public static class LoginAttempt extends BaseDomainModel {
...
+                return selectedInstitution.get();
+            }
+        }
+
+        return this.institutions.stream()
+                .filter(Institution::isActive)
+                .findFirst()
+                .orElseThrow(() -> new AysUserHasNoActiveInstitutionException(this.id));
+    }
+
```
Comment (agitrubard) [inline src/main/java/org/ays/auth/model/AysUser.java]: Buradaki `isActive` filtresi iki yerde tekrar ediyor gibi, bir kere `activeInstitutions` diye ayrı bir listeye çekip onun üzerinden gidebiliriz belki, biraz daha temiz durur. Bir de metodun ismi `resolveActiveInstitution` ama aslında önce son seçileni bulmaya çalışıyor, o yüzden `findSelectedActiveInstitution` gibi bir isim yaptığı işi daha iyi anlatabilir. Son olarak `String` + `orElse(null)` + null kontrolü yerine `Optional` akışında bırakırsak `orElse(null)`'dan da kurtulmuş oluruz

javadoc'u da bu koda paralel olacak şekilde güncelleyebiliriz

```suggestion
    public Institution findSelectedActiveInstitution() {

        if (CollectionUtils.isEmpty(this.institutions)) {
            throw new AysUserHasNoActiveInstitutionException(this.id);
        }

        final Optional<String> lastSelectedInstitutionId = Optional.ofNullable(this.loginAttempt)
                .map(LoginAttempt::getLastSelectedInstitutionId);

        final List<Institution> activeInstitutions = this.institutions.stream()
                .filter(Institution::isActive)
                .toList();

        if (lastSelectedInstitutionId.isPresent()) {

            final Optional<Institution> selectedInstitution = activeInstitutions.stream()
                    .filter(activeInstitution -> activeInstitution.getId().equals(lastSelectedInstitutionId.get()))
                    .findFirst();

            if (selectedInstitution.isPresent()) {
                return selectedInstitution.get();
            }
        }

        return activeInstitutions.stream()
                .findFirst()
                .orElseThrow(() -> new AysUserHasNoActiveInstitutionException(this.id));
    }
```

File: src/main/java/org/ays/auth/service/impl/AysAuthServiceImpl.java [hunk 10]
```diff
@@ -89,6 +89,10 @@ public AysToken authenticate(final AysLoginRequest loginRequest) {
                             user.setLoginAttempt(AysUser.LoginAttempt.builder().build());
                             user.getLoginAttempt().success();
                         });
+
+        final Institution activeInstitution = user.resolveActiveInstitution();
```
Comment (agitrubard) [inline src/main/java/org/ays/auth/service/impl/AysAuthServiceImpl.java]:
```suggestion
        final Institution selectedActiveInstitution = user.findSelectedActiveInstitution();
```

File: src/test/java/org/ays/auth/service/impl/AysAuthServiceImplTest.java [hunk 11]
```diff
@@ -201,6 +202,142 @@ void givenValidLoginRequest_whenUserAuthenticatedForLanding_thenReturnAysToken()
...
+
+        Assertions.assertEquals(mockUserToken, token);
+        Assertions.assertEquals(
+                activeInstitution.getId(),
+                mockUser.getLoginAttempt().getLastSelectedInstitutionId()
+        );
+
+        // Verify
+        Mockito.verify(userSavePort, Mockito.times(1))
+                .save(Mockito.any(AysUser.class));
```
Comment (agitrubard) [inline src/test/java/org/ays/auth/service/impl/AysAuthServiceImplTest.java:353]: Bu kısımda mockito when ile tanımladığımız metotları da verify adımına ekleyebilir miyiz?

File: src/test/java/org/ays/auth/service/impl/AysAuthServiceImplTest.java [hunk 12]
```diff
@@ -201,6 +202,142 @@ void givenValidLoginRequest_whenUserAuthenticatedForLanding_thenReturnAysToken()
...
+
+        Assertions.assertEquals(mockUserToken, token);
+        Assertions.assertEquals(
+                lastSelectedInstitution.getId(),
+                mockUser.getLoginAttempt().getLastSelectedInstitutionId()
+        );
+
+        // Verify
+        Mockito.verify(userSavePort, Mockito.times(1))
+                .save(Mockito.any(AysUser.class));
```
Comment (agitrubard) [inline src/test/java/org/ays/auth/service/impl/AysAuthServiceImplTest.java:275]: Bu kısımda mockito when ile tanımladığımız metotları da verify adımına ekleyebilir miyiz?

**Review by agitrubard — CHANGES_REQUESTED** [review]
Yazdığım yorumlara ek olarak end-to-end test ihtiyacımız da var diye düşünüyorum. Özellikle verinin doğru şekilde ilgili senaryolarda veri tabanına yazıldığından emin olmamız daha sağlıklı olur. Ek olarak edge case'lerdeki unit testleri de ekleyebiliriz, coverage kısmından kontrol edip hangi yerlere hit etmediğini kontrol edebilirsin

**Review by agitrubard — APPROVED** [review]
